File: UMPTest/UMPTest/InteropWithCpp/TestInteropWithCpp.swift
//
//  TestInteropWithCpp.swift
//  UMPTest
//
//  Created by 龚文 on 10/23/23.
//

import Foundation

public class TestInteropWithCpp {
    /**
     * 验证在Framework中 Swift-C++互调实现
     */
    public static func main() {
        var person = Person2("commercial1", 666)
        let result = person.callFromSwift("param1")
        print("Swift-Cpp-interop [Swfit call C++]  =>  Result: \(result)")

        let result2 = String(cString: callFromSwift("param3"))
        print("Swift-Cpp-interop Direct Call [Swift call C++]   =>  Result: \(result2)")
    }

    public static func callFromCpp(param: String) -> String {
        print("Swift-Cpp-interop Direct Call [C++ call Swift]   =>  Param: \(param)")
        return "done " + param
    }
}

public class SwiftPerson {
    public var name: String
    public var age: Int

    public init(name: String, age: Int) {
        self.name = name
        self.age = age
    }

    public func callFromCpp(param: String) -> String {
        print("Swift-Cpp-interop [C++ call Swfit]  =>  Name: \(name), Age: \(age), Param: \(param)")
        return "done " + param
    }
}

File: UMPTest/UMPTest/InteropWithCpp/TestInteropWithCpp.cpp
//
//  TestInteropWithCpp.cpp
//  UMPTest
//
//  Created by 龚文 on 10/23/23.
//

#include <iostream>
#include <UMPTest/UMPTest-Swift.h>
#include "TestInteropWithCpp.hpp"


Person2::Person2(std::string name, int age) {
    m_name = name;
    m_age = age;
}

std::string Person2::callFromSwift(std::string param) {
    std::cout << "Swift-Cpp-interop [Swfit call C++]  =>  " << "Name: " << m_name << ", Age: " << m_age << ", Param: " << param << std::endl;

    UMPTest::SwiftPerson swiftPerson = UMPTest::SwiftPerson::init("commercial2", 666);
    std::string result = swiftPerson.callFromCpp("param2");
    std::cout << "Swift-Cpp-interop [C++ call Swift]  =>  " << "Result: " << result << std::endl;

    return "done " + param;
}

char * callFromSwift(const char *param) {
    std::cout << "Swift-Cpp-interop Direct Call [Swift call C++]   =>  " << "Param: " << param << std::endl;

    std::string result = UMPTest::TestInteropWithCpp::callFromCpp("param4");
    std::cout << "Swift-Cpp-interop Direct Call [C++ call Swift]   =>  " << "Result: " << result << std::endl;

    char *result2 = strdup("done param3");

    // free(result2);
    return result2;
}
